### ISSUE TITLE
Pruning off NaN values in the gradient graph still produces NaN gradients.

### ISSUE THREAD

**Issue by kaniblu** [pytorch/pytorch#23156]
## 🐛 Bug

This is a niche bug, but it might cause troubles in advanced users who like to use masking to filter out NaN losses. Simply put, when NaN losses are masked out using `masked_fill`, performing `backward` on the sum of the losses should produce valid gradients (assuming that the gradient graph is smooth everywhere except for the masked losses). When 

## To Reproduce

Define the environment as follows. We will be backpropagating gradients to a very simple `nn.Linear` layer.

```python
import torch
import torch.nn as nn
x = torch.ones(3, 4)
linear = nn.Linear(4, 2)
def has_err(x):
    return bool(((x != x) | (x == float("inf")) | (x == float("-inf"))).any().item())
```

Performing a forward inference on the linear layer and propagating gradients from there works as intended if no `nan` is involved.

```python
linear(x).sum().backward()
print(linear.weight.grad)
# tensor([[3., 3., 3., 3.],
#         [3., 3., 3., 3.]])
assert not has_err(linear.weight.grad)
```

Suppose that `x` has `nan` in one of its rows.

```python
x[2].fill_(float("nan"))
print(linear(x))
# tensor([[ 0.0570, -1.0066],
#         [ 0.0570, -1.0066],
#         [    nan,     nan]], grad_fn=<AddmmBackward>)
```

Performing backwards computation on the sum of the matrix above should
produce `nan` gradients. This is expected.

```python
linear.zero_grad()
linear(x).sum().backward()
print(linear.weight.grad)
# tensor([[nan, nan, nan, nan],
#         [nan, nan, nan, nan]])
assert has_err(linear.weight.grad)
```

However, using `masked_fill` or  index slicing on the leaf nodes to mask out problematic gradient graphs should not.

```python
linear.zero_grad()
linear(x)[:2].sum().backward()
assert not has_err(linear.weight.grad)  # AssertionError raised
```
We can check whether masking out final losses works in cases where no `nan` is involved.

```python
x = torch.ones(3, 4)
x[2].fill_(0.5)
linear.zero_grad()
linear(x).sum().backward()
print(linear.weight.grad)
# without masking:
# tensor([[2.5000, 2.5000, 2.5000, 2.5000],
#         [2.5000, 2.5000, 2.5000, 2.5000]])
linear.zero_grad()
linear(x)[2:].sum().backward()
print(linear.weight.grad)
# with masking:
# tensor([[2., 2., 2., 2.],
#         [2., 2., 2., 2.]])
```

The accumulated gradient in the latter case is less than the former because index slicing (`[2:]`) on the computation graph prevented the third row in `x` from affecting the gradient computation.

## Environment

Collecting environment information...
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.14.5
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy==1.16.3
[pip] torch==1.0.0
[conda] torch                     1.0.0                     <pip>

## Additional context

None.

**Comment by colesbury** [comment]
This isn't a bug. The gradient formula for linear is `weight.grad = x @ grad_output`. You have NaNs in `x`. It doesn't matter that grad_output contains zeros. NaN * 0 is still NaN with IEEE-754 floating point arithmetic.

**Comment by fmassa** [comment]
This is a recurrent issue, see #15506 #12986 as an example

Quoting @ppwwyyxx from one of our previous discussions

> The fundamental issue is that autograd does not distinguish "no gradient" and "zero gradient".

Given that this has happened very often in the past, I'm re-opening the issue and tagging it as needs discussion. One possible way of solving this would be to carry an additional (sparse?) tensor that can distinguish between "no gradient" and "zero gradient".

**Comment by kaniblu** [comment]
@fmassa This indeed seems like a popular issue that keeps coming back (more examples: #15131 #17455). After reading several closed discussions, I am not sure if it's possible to come up with an elegant solution that can substitute masking with arithmetic multiplication. Passing around sparse flags might be a solution, but it seems like awful lot of work compared to the severity of the bug and could introduce unforeseen consequences like a can of worm.

**Comment by izdeby** [comment]
We should add explanation for this behavior in out FAQ

**Comment by carlosgmartin** [comment]
What's the best way to compute the entropy of a vector, i.e. `-(p*p.log()).sum(-1)`, when some of the `p`s are occasionally 0, in such a way that one doesn't get `nan` gradients? Because of this issue, `-torch.where(p > 0, p*p.log(), torch.zeros(p.shape)).sum(-1)` doesn't work since it can work in the forward pass but return `nan`s in the backward pass.

**Comment by kaniblu** [comment]
@carlosgmartin For now, the best way is to apply the mask before the subjecting the operand to the operations that results in NaNs. So for your case,

```
p = p.masked_fill(p <= 0, 1)
p.mul(p.log()).sum(-1)
```

this should do the trick.

**Comment by jonasrauber** [comment]
This can happen in all kinds of situations and should be fixed https://github.com/pytorch/pytorch/issues/36923

**Comment by kaniblu** [comment]
Thank you for bringing this to the attention. An overall maintenance would be desirable at this point.

________________________________

**Comment by EricCousineau-TRI** [comment]
FWIW I'm guessing that I ran into this too here:
https://gist.github.com/EricCousineau-TRI/4d9ec78adf4b566b6d1ce7ab4b643670#file-torch_mask_gradient_nonfinite-py-L98-L108

In a simplified case, I am doing an element-wise multiplication of two vectors, where the constant vector may have an entry of `inf` (for depth images). Basically:
```py
def get_vars():
    a = torch.tensor([inf, 1.0])
    b = torch.tensor([2.0, 3.0], requires_grad=True)
    return a, b

a, b = get_vars()
y1 = a * b
J1 = jacobian(b, y1)  # Values will have Inf, gradients will have Inf and NaNs.
# v2
a, b = get_vars()
y2 = a * b
y2[~torch.isfinite(y2)] = 0.0  # This will zero out values, but gradients will have NaNs.
J2 = jacobian(b, y2)
# v3
a, b = get_vars()
a[~torch.isfinite(a)] = 0.0  # This should be OK.
y3 = a * b
J3 = jacobian(b, y3)
```

I think I understand all of this behavior except for these two cases. Here's the output for those two cases:
https://gist.github.com/EricCousineau-TRI/4d9ec78adf4b566b6d1ce7ab4b643670#file-torch_mask_gradient_nonfinite-py-L98-L108

Can I ask if someone can confirm if this is indeed the same issue?
I am confused as to why `J[0, 1]` is `nan` for `v2` (post-masking) and `v3` (pre-masking), and why `J[0, 0]` is `nan` for `v2` (post-masking).

I'm happy to make a separate issue if that simplifies things.

**Comment by orsharir** [comment]
It seems like this issue was already fixed at some point. The following snippet taken from OP that was supposed to raise an assertion, but now passes without any when tested on Colab (on both cpu and gpu).
```python
import torch
import torch.nn as nn
x = torch.ones(3, 4, device='cuda')
linear = nn.Linear(4, 2).to(torch.device('cuda'))
def has_err(x):
    return bool(((x != x) | (x == float("inf")) | (x == float("-inf"))).any().item())

linear.zero_grad()
linear(x)[:2].sum().backward()
assert not has_err(linear.weight.grad)  # AssertionError raised
```

**Comment by EricCousineau-TRI** [comment]
@orsharir Can you post a Gist with `pip freeze` in your Colab env?

I tried your rendition of the code on my machine (Linux Ubuntu 18.04, CPython 3.6.9, Titan RTX, `torch==1.7.1`), but did not get an assertion error. I also have the same behavior on my above snippet.

**Comment by EricCousineau-TRI** [comment]
Hm... to that point, I can't repro OP's error (using Or's consolidated code) on torch 1.0.0, 1.4.0, or 1.7.1...

**Comment by TobbysGitHub** [comment]
I am testing a new activation function named which is like:
<img src="https://latex.codecogs.com/svg.latex?{\color{Blue}&space;\begin{cases}&space;\frac{1}{2-x}&space;&&space;\text{&space;if&space;}&space;x<1&space;\\&space;x&&space;\text{&space;if&space;}&space;x>=1&space;\end{cases}}" title="{\color{Blue} \begin{cases} \frac{1}{2-x} & \text{ if } x<1 \\ x& \text{ if } x>=1 \end{cases}}" />, 
and I use the following code:
`torch.nan_to_num((x < 1) * (1 / (2 - x))) + (x >= 1) * x`.
After 40511 iterations, it happens that one element of x equals 2, and so I get a nan~~.

**Comment by carlosgmartin** [comment]
@fmassa @albanD Couldn't the [backward pass](https://pytorch.org/tutorials/beginner/examples_autograd/two_layer_net_custom_function.html) for `torch.where`, `torch.masked_fill`, etc. be implemented as true conditionals, rather than multiplication with binary masks? Wouldn't that solve the problem?

**Comment by albanD** [comment]
Hi,

As mentioned in other issues, the problem is not with the where op. It produces 0s as expected. It is the function just before that that transforms this 0 into NaN. And the only way to solve this would be to see these two functions as a single one where we could set a specific formula (which you can do with a custom Function if you want to), otherwise, you can try to use things like https://github.com/cpuhrsch/maskedtensor that properly differentiates 0s from not-present values.

**Comment by soulitzer** [comment]
Duplicate of https://github.com/pytorch/pytorch/issues/4132

**Comment by profPlum** [comment]
> We should add explanation for this behavior in out FAQ

Don't explain problems, fix them.
